shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/EpochBoundary.hs

```diff
@@ -79,7 +80,19 @@ getStakeHK _ = Nothing
 
 aggregateOuts :: Crypto crypto => UTxO crypto -> Map (Addr crypto) Coin
 aggregateOuts (UTxO u) =
-  Map.fromListWith (+) (map (\(_, TxOut a c) -> (a, c)) $ Map.toList u)
+  -- Map.fromListWith (+) (map (\(_, TxOut a c) -> (a, c)) $ Map.toList u)
+  Map.foldr accum Map.empty u
+    where accum (TxOut a c) ans = Map.insertWith (+) a c ans
```

Comment:
I see you're no longer using `aggregateOuts` at all, so the comment below may no longer be relevant:

When looking at the [profiteur](https://hackage.haskell.org/package/profiteur) output (attached) of `cardano-node` doing a full sync with MC1, you can drill down on the top left corner until you get to `aggregateOuts` (38.6% of the total time), when you go deeper, you'll see that 24.9% of the total time is spent in `deserialiseAddr`. That's because pattern matching on a `TxOut`, which is a pattern synonym for a `TxOutCompact`, deserialises the address (from a `ShortByteString`):

```
pattern TxOut ::
  Crypto crypto =>
  Addr crypto ->
  Coin ->
  TxOut crypto
pattern TxOut addr coin <-
  (viewCompactTxOut -> (addr, coin))
  where
    TxOut addr (Coin coin) =
      TxOutCompact (BSS.toShort $ serialiseAddr addr) (fromIntegral coin)
```
Idea: wouldn't it be more efficient to keep using the `ShortByteString`s as keys in the map while doing `Map.insertWith (+)` and afterwards converting the keys in the final map from `ShortByteString`s to `Addr`s?

[cardano-node.prof.html.gz](https://github.com/input-output-hk/cardano-ledger-specs/files/4970644/cardano-node.prof.html.gz)

Reply:
This is an excellent observation. But the problem really has two components. The observation is that
(TxOut addr (Coin coin)) is a pattern and the pattern has to deserializes the whole address,
to get the part of the address that holds the credential. It turns out there is a second problem that
that makes this really slow. The 'addr' is stored as a ShortByteString, and that has to be converted 
to a ByteString before it can be deserialized. The solution is to deserialize directly from the ShortByteString, and only look
at the Credential part of the address. We implemented this, and created some benchmarks, The
new code is 10x faster than the old code.